### PR TITLE
EOS-18207: bucket metadata cache - part 2

### PR DESCRIPTION
**Pull request by evgeniy-brazhnikov**
Signed-off-by: evgeniy-brazhnikov <evgeniy.brazhnikov@seagate.com>

### CODE REVIEW

File: server/s3_bucket_metadata_cache.cc [hunk 1]
```diff
@@ -45,53 +45,61 @@ S3BucketMetadataCache::Item::Item(const S3BucketMetadata& src)
 }
 
 S3BucketMetadataCache::Item::~Item() {
-  assert(can_remove());
-  assert(fetch_waiters.empty());
-  assert(!on_changed);
+  if (!can_remove()) {
+    s3_log(S3_LOG_FATAL, nullptr, "Motr operation is in progress");
```
Comment (knrajnambiar76) [inline server/s3_bucket_metadata_cache.cc:49]: which scenario this case can arise ?

Comment (evgeniy-brazhnikov) [reply server/s3_bucket_metadata_cache.cc:49]: It's for detecting possible logic errors in future

File: server/s3_bucket_metadata_cache.h [hunk 2]
```diff
@@ -51,24 +54,29 @@ class S3BucketMetadataCache {
 
   virtual ~S3BucketMetadataCache();
 
+  static S3BucketMetadataCache* get_instance();
+
   void disable() { disabled = true; }
   void enable() { disabled = false; }
 
   using FetchHanderType =
```
Comment (knrajnambiar76) [inline server/s3_bucket_metadata_cache.h:62]: have comments on it and some other places also

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_bucket_metadata_cache.h (hunk 2 above)
Comment (evgeniy-brazhnikov) [reply server/s3_bucket_metadata_cache.h:62]: Will add.

File: server/s3_bucket_metadata_cache.cc (hunk 1 above)
Comment (amitwac2608) [reply server/s3_bucket_metadata_cache.cc:49]: we need to give more info in the log message, to figure out what went wrong... 

_"Motr operation is in progress" is not enough._

Give info like _"during X, Y happened, which is not expected. Please check or fix Z."_

Comment (knrajnambiar76) [reply server/s3_bucket_metadata_cache.cc:49]: More message can be given, which is good

better not to say to "fix Z" unless we know exactly how that can be fixed.

**Review by amitwac2608 — APPROVED** [review]

**Pull request merged**